compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIECanonicalizeDoublyStridedOp.cpp

```diff
@@ -0,0 +1,111 @@
+  SmallVector<OpFoldResult> targetStrides = op.getTargetMixedStrides();
+  SmallVector<OpFoldResult> newSourceOffsets, newSourceSizes, newSourceStrides,
+      newTargetOffsets, newTargetSizes, newTargetStrides;
+  LogicalResult sourceRes =
+      foldLinearDims(op.getContext(), sourceOffsets, sourceSizes, sourceStrides,
+                     newSourceOffsets, newSourceSizes, newSourceStrides);
+  LogicalResult targetRes =
+      foldLinearDims(op.getContext(), targetOffsets, targetSizes, targetStrides,
+                     newTargetOffsets, newTargetSizes, newTargetStrides);
+  if (failed(sourceRes) && failed(targetRes)) {
+    return failure();
+  }
+
+  rewriter.setInsertionPointAfter(op);
+  auto newDoublyStridedOp = op.createDoublyStridedOp(
+      rewriter, newTargetOffsets, newTargetSizes, newTargetStrides,
+      newSourceOffsets, newSourceSizes, newSourceStrides);
+  rewriter.replaceOp(op, newDoublyStridedOp.getOperation());
+  return success();
+}
+
+/// Fold single dimension linear accesses and make them implicit.
+LogicalResult foldDmaOpSingleDims(RewriterBase &rewriter,
+                                  AMDAIE::DoublyStridedOpInterface op) {
+  OpBuilder::InsertionGuard guard(rewriter);
+  SmallVector<OpFoldResult> sourceOffsets = op.getSourceOffsets();
+  SmallVector<OpFoldResult> sourceSizes = op.getSourceSizes();
+  SmallVector<OpFoldResult> sourceStrides = op.getSourceStrides();
+  SmallVector<OpFoldResult> targetOffsets = op.getTargetOffsets();
+  SmallVector<OpFoldResult> targetSizes = op.getTargetSizes();
+  SmallVector<OpFoldResult> targetStrides = op.getTargetStrides();
+  LogicalResult sourceRes =
+      foldSingleDim(sourceOffsets, sourceSizes, sourceStrides);
+  LogicalResult targetRes =
+      foldSingleDim(targetOffsets, targetSizes, targetStrides);
+  if (failed(sourceRes) && failed(targetRes)) {
+    return failure();
+  }
+
+  rewriter.setInsertionPointAfter(op);
+  auto newDoublyStridedOp = op.createDoublyStridedOp(
+      rewriter, targetOffsets, targetSizes, targetStrides,
+      sourceOffsets, sourceSizes, sourceStrides);
+  rewriter.replaceOp(op, newDoublyStridedOp.getOperation());
+  return success();
+}
+
+class AMDAIECanonicalizeDoublyStridedOpPass
+    : public impl::AMDAIECanonicalizeDoublyStridedOpBase<
+          AMDAIECanonicalizeDoublyStridedOpPass> {
+ public:
+  void getDependentDialects(DialectRegistry &registry) const override {
+    registry.insert<AMDAIEDialect>();
+  }
+
+  AMDAIECanonicalizeDoublyStridedOpPass() = default;
+  AMDAIECanonicalizeDoublyStridedOpPass(
+      const AMDAIECanonicalizeDoublyStridedOpPass &pass){};
+  void runOnOperation() override;
+};
+
+void AMDAIECanonicalizeDoublyStridedOpPass::runOnOperation() {
+  Operation *parentOp = getOperation();
+  IRRewriter rewriter(parentOp->getContext());
+
+  // Fold linear dimensions within a DMA op.
+  parentOp->walk([&](AMDAIE::DoublyStridedOpInterface dmaOp) {
+    (void)foldDmaOpLinearDims(rewriter, dmaOp);
+  });
+
```

Comment:
IIRC, in the existing flow, the unit dim canonicalization was helping even in permuted cases where a dim was unit so could be dropped. Is something expected to be different in this flow that you anticipate not needing it or just haven't seen a use case yet?

Reply:
I haven't seen this use case yet. Let me have a look.

Reply:
I added the unit dimension folding as well now: https://github.com/nod-ai/iree-amd-aie/pull/327/files#diff-9a02ada48f4391792b10c5eab979fe266d5c376ffdbc0bc52d5e34e3bf910212R130 + added specific testcases for this as well.